src/Compilers/CSharp/Test/Semantic/Semantics/StackAllocSpanExpressionsTests.cs

```diff
@@ -0,0 +1,488 @@
+{
+    void M()
+    {
+        double x = stackalloc int[10];          // implicit
+        short y = (short)stackalloc int[10];    // explicit
+    }
+}", TestOptions.UnsafeReleaseDll).VerifyDiagnostics(
+                // (6,20): error CS8520: Conversion of a stackalloc expression of type 'int' to type 'double' is not possible.
+                //         double x = stackalloc int[10];          // implicit
+                Diagnostic(ErrorCode.ERR_StackAllocConversionNotPossible, "stackalloc int[10]").WithArguments("int", "double").WithLocation(6, 20),
+                // (7,19): error CS8520: Conversion of a stackalloc expression of type 'int' to type 'short' is not possible.
+                //         short y = (short)stackalloc int[10];    // explicit
+                Diagnostic(ErrorCode.ERR_StackAllocConversionNotPossible, "(short)stackalloc int[10]").WithArguments("int", "short").WithLocation(7, 19));
+        }
+
+        [Fact]
+        public void MissingSpanType()
+        {
+            CreateStandardCompilation(@"
+class Test
+{
+    void M()
+    {
+        Span<int> a = stackalloc int [10];
+    }
+}").VerifyDiagnostics(
+                // (6,9): error CS0246: The type or namespace name 'Span<>' could not be found (are you missing a using directive or an assembly reference?)
+                //         Span<int> a = stackalloc int [10];
+                Diagnostic(ErrorCode.ERR_SingleTypeNameNotFound, "Span<int>").WithArguments("Span<>").WithLocation(6, 9),
+                // (6,23): error CS0518: Predefined type 'System.Span`1' is not defined or imported
+                //         Span<int> a = stackalloc int [10];
+                Diagnostic(ErrorCode.ERR_PredefinedTypeNotFound, "stackalloc int [10]").WithArguments("System.Span`1").WithLocation(6, 23));
+        }
+
+        [Fact]
+        public void MissingSpanConstructor()
+        {
+            CreateStandardCompilation(@"
+namespace System
+{
+    ref struct Span<T>
+    {
+    }
+    class Test
+    {
+        void M()
+        {
+            Span<int> a = stackalloc int [10];
+        }
+    }
+}").VerifyDiagnostics(
+                // (11,27): error CS0656: Missing compiler required member 'System.Span`1..ctor'
+                //             Span<int> a = stackalloc int [10];
+                Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "stackalloc int [10]").WithArguments("System.Span`1", ".ctor").WithLocation(11, 27));
+        }
+
+        [Fact]
+        public void ConditionalExpressionOnSpan_NonConvertible()
+        {
+            CreateCompilationWithMscorlibAndSpan(@"
+class Test
+{
+    void M()
+    {
+        var x = true ? stackalloc int [10] : stackalloc int [5];
+    }
+}", TestOptions.UnsafeReleaseDll).VerifyDiagnostics(
+                // (6,17): error CS0173: Type of conditional expression cannot be determined because there is no implicit conversion between 'stackalloc expression of type 'int'' and 'stackalloc expression of type 'int''
```

Comment:
I understand why this error is here, but perhaps we should specialize on stackalloc conversions and have a better message?  

It's not immediately obvious why "stackalloc expression of type 'int'` can't be converted to `stackalloc expression of type 'int'`.